File: 11-container-with-most-water/container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int start = 0, end = height.size()-1, res = 0;
        while (start < end) {
            int width = end-start;
            int h = min(height[start], height[end]);
            int temp = width * h;
            res = max(res, temp);
            if (height[start] < height[end]) start++;
            else end--;
        }
        return res;
    }
};

File: 11-container-with-most-water/container-with-most-water.py
class Solution(object):
    def maxArea(self, height):
        i, j = 0, len(height) - 1
        best = 0
        while i < j:
            h = min(height[i], height[j])
            best = max(best, h * (j - i))
            # Greedy rule: move the shorter wall inward
            if height[i] < height[j]:
                i += 1
            else:
                j -= 1
        return best
